dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/EPersonRestRepository.java

```diff
@@ -82,8 +124,63 @@ protected EPersonRest createAndReturn(Context context)
         } catch (SQLException e) {
             throw new RuntimeException(e.getMessage(), e);
         }
+        return eperson;
+    }
+
+    private EPersonRest createAndReturn(Context context, EPersonRest epersonRest, String token)
+        throws AuthorizeException, SQLException {
+        RegistrationData registrationData = registrationDataService.findByToken(context, token);
+        if (registrationData == null) {
+            throw new DSpaceBadRequestException("The token given as parameter: " + token + " does not exist" +
+                                                " in the database");
+        }
+        if (es.findByEmail(context, registrationData.getEmail()) != null) {
+            throw new DSpaceBadRequestException("The token given already contains an email address that resolves" +
+                                                "to an eperson");
+        }
+        String emailFromJson = epersonRest.getEmail();
+        if (StringUtils.isNotBlank(emailFromJson)) {
+            if (!StringUtils.equalsIgnoreCase(registrationData.getEmail(), emailFromJson)) {
+                throw new DSpaceBadRequestException("The email resulting from the token does not match the email given"
+                                                        + " in the json body. Email from token: " +
+                                                    registrationData.getEmail() + " email from the json body: "
+                                                    + emailFromJson);
+            }
+        }
+        if (epersonRest.isSelfRegistered() != null && !epersonRest.isSelfRegistered()) {
+            throw new DSpaceBadRequestException("The self registered property cannot be set to false using this method"
+                                                    + " with a token");
+        }
+        checkRequiredProperties(epersonRest);
+        if (!AuthorizeUtil.authorizeNewAccountRegistration(context, requestService
+            .getCurrentRequest().getHttpServletRequest())) {
+            throw new DSpaceBadRequestException(
+                "Registration is disabled, you are not authorized to create a new Authorization");
```

Comment:
Again, this should be moved to the top of the method.  No need to do all the checks above if registration is disabled altogether.

Reply:
This has been adjusted